### PR TITLE
Dev fix high order derivative

### PR DESCRIPTION
**Pull request by hjchen2**
Fix https://github.com/Oneflow-Inc/OneTeam/issues/1181

增加一个新的AutogradCapturedTensor，内部对FunctionNode使用弱引用，避免造成输出Tensor和FunctionNode的循环引用，同时使得高阶导的构图可以正确回溯到输入的grad function node。

### CODE REVIEW

**Review by wyg1997 — APPROVED** [review]

File: oneflow/core/framework/tensor.h [hunk 1]
```diff
@@ -267,107 +267,101 @@ class TensorIf : public Tensor {
...
-  bool is_consistent() const override { return tensor_->is_consistent(); }
-  bool is_local() const override { return tensor_->is_local(); }
-  bool is_lazy() const override { return tensor_->is_lazy(); }
-  bool is_eager() const override { return tensor_->is_eager(); }
-  const TensorMeta& tensor_meta() const override { return tensor_->tensor_meta(); }
-  Maybe<Symbol<ConsistentTensorMeta>> consistent_tensor_meta() const override {
+  ProxyTensor(const std::shared_ptr<Tensor>& tensor) : tensor_(tensor) {}
+  virtual ~ProxyTensor() = default;
+
+  virtual const std::shared_ptr<const Shape>& shape() const { return tensor_->shape(); }
```
Comment (daquexian) [inline oneflow/core/framework/tensor.h:283]: virtual 函数也可以标上 override

**Review by daquexian — APPROVED** [review]

File: oneflow/core/framework/tensor.h (hunk 1 above)
Comment (hjchen2) [reply oneflow/core/framework/tensor.h]: 已加上

File: oneflow/core/framework/tensor.h [hunk 2]
```diff
@@ -376,24 +380,35 @@ class Parameter final : public TensorIf<Parameter> {
...
 
- private:
+ protected:
   std::shared_ptr<Tensor> tensor_;
 };
 
+class Parameter final : public ProxyTensor<Parameter> {
+ public:
+  Parameter(const std::shared_ptr<Tensor>& tensor, bool requires_grad)
+      : ProxyTensor<Parameter>(tensor->detach().GetPtrOrThrow()) {
```
Comment (daquexian) [inline oneflow/core/framework/tensor.h:404]: 这里有一个问题是可能 `tensor` 本身还是一个 parameter，这种情况 AsMirroredTensor 就会报错，看是不是可以把 ProxyTensor 的 AsMirroredTensor 实现成 `return tensor_->AsMirroredTensor();`

Comment (daquexian) [reply oneflow/core/framework/tensor.h:404]: 哦哦没事了，看到了 ProxyTensor 的 detach() 方法

Comment (hjchen2) [reply oneflow/core/framework/tensor.h:404]: 不会，ProxyTensor<Parameter>构造拿的是`tensor->detach()`的结果，即使tensor是parameter，detach之后结果也是普通tensor

File: oneflow/core/autograd/autograd_captured_tensor.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+#include "oneflow/core/framework/tensor.h"
+#include "oneflow/core/functional/functional.h"
+
+namespace oneflow {
+namespace one {
+
+class AutogradCapturedTensor final : public ProxyTensor<AutogradCapturedTensor> {
+ public:
+  AutogradCapturedTensor(const std::shared_ptr<Tensor>& tensor)
```
Comment (lixinqi) [inline oneflow/core/autograd/autograd_captured_tensor.h:27]: explicit

Comment (lixinqi) [reply oneflow/core/autograd/autograd_captured_tensor.h:27]: 可不可以把构造置为private。
然后提供 static Maybe<AutogradCapturedTensor> New(const std::shared_ptr<Tensor>& tensor)方法。
这样的好处是，把绝大部分初始化逻辑和检查放置在单独的Init函数里，避免直接使用GetPtrOrThrow以及CHECK。

File: oneflow/core/framework/tensor.h [hunk 4]
```diff
@@ -267,107 +267,111 @@ class TensorIf : public Tensor {
...
+  std::shared_ptr<FunctionNode> mut_grad_fn_node() override { return grad_fn_node_; }
 
  protected:
   TensorIf() = default;
   std::shared_ptr<FunctionNode> grad_fn_node_;
 };
 
-class Parameter final : public TensorIf<Parameter> {
+template<typename DerivedT>
+class ProxyTensor : public TensorIf<DerivedT> {
```
Comment (lixinqi) [inline oneflow/core/framework/tensor.h:278]: 其实这个名字非常不好起。我大概能猜测出来ProxyTensor内会包含一个Tensor指针。看上去是为了复用代码而提供的基类。

类的继承关系中，公共基类的子类应该有很强的语义关联。AutogradCapturedTensor和Parameter同时继承了ProxyTensor，但其实它们没有太直接的关系，只是ProxyTensor提炼了它们绝大部分共同的方法。

我不太确定这种方式是否好，可以请俊丞建浩看看。

File: oneflow/core/autograd/autograd_captured_tensor.h (hunk 3 above)
Comment (hjchen2) [reply oneflow/core/autograd/autograd_captured_tensor.h:27]: 可以，我改一下

File: oneflow/core/framework/tensor.h (hunk 4 above)
Comment (hjchen2) [reply oneflow/core/framework/tensor.h:278]: 本来也想用TensorInterface之类的来命名，但TensorIf已经是一个基类了。ProxyTensor其实相比TensorInterface更贴近这里的意图，如果觉得Parameter和AutogradCapturedTensor不应该有共同基类，那其实就是不应该让AutogradCapturedTensor是一个Tensor，只要它是Tensor，那Parameter和AutogradCapturedTensor就会有一个基类TensorIf。

Comment (hjchen2) [reply oneflow/core/framework/tensor.h:278]: 如果AutogradCapturedTensor不是一个Tensor，那grad_fn就需要同时捕获detached tensor（没有持有grad_fn）和grad_fn的弱引用，计算一阶导时autograd engine就得拿这个捕获的detached tensor和grad_fn的弱引用去构二阶导的后向图，这个改变对现在autograd engine的逻辑是破坏性的，没有现在这样方便和自然。

Comment (lixinqi) [reply oneflow/core/framework/tensor.h:278]: 嗯。ProxyTensor看起来比其他方案还是稍微好点。

File: oneflow/core/autograd/autograd_captured_tensor.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+ private:
+  explicit AutogradCapturedTensor(const std::shared_ptr<Tensor>& tensor)
+      : ProxyTensor<AutogradCapturedTensor>(tensor->detach().GetPtrOrThrow()) {
+    this->tensor_->set_autograd_meta(tensor->mut_autograd_meta());
+    if (tensor->requires_grad()) {
+      CHECK(tensor->grad_fn_node()) << "The grad function node is expected for the captured tensor "
+                                       "which requires_grad is True";
+    }
+    grad_fn_node_ = tensor->mut_grad_fn_node();
```
Comment (lixinqi) [inline oneflow/core/autograd/autograd_captured_tensor.h:51]: 可不可以把这一段代码放置在Maybe<void> Init()函数中。这样CHECK就可以替换为CHECK_OR_RETURN

File: oneflow/core/autograd/autograd_captured_tensor.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  std::shared_ptr<Tensor> contiguous() const override {
+    const auto& tensor = std::const_pointer_cast<Tensor>(shared_from_this());
+    if (tensor_->is_contiguous()) { return tensor; }
+    return CHECK_JUST(functional::ToContiguous(tensor));
+  }
+
+ private:
+  explicit AutogradCapturedTensor(const std::shared_ptr<Tensor>& tensor)
+      : ProxyTensor<AutogradCapturedTensor>(tensor->detach().GetPtrOrThrow()) {
```
Comment (lixinqi) [inline oneflow/core/autograd/autograd_captured_tensor.h:45]: 这个GetPtrOrThrow应该是可以去掉的，让MakeTensor直接传一个tensor->detach()的值。

File: oneflow/core/autograd/autograd_captured_tensor.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+#include "oneflow/core/framework/tensor.h"
+#include "oneflow/core/functional/functional.h"
+
+namespace oneflow {
+namespace one {
+
+class AutogradCapturedTensor final : public ProxyTensor<AutogradCapturedTensor> {
+ public:
+  static Maybe<AutogradCapturedTensor> MakeTensor(const std::shared_ptr<Tensor>& tensor) {
+    return std::shared_ptr<AutogradCapturedTensor>(new AutogradCapturedTensor(tensor));
```
Comment (lixinqi) [inline oneflow/core/autograd/autograd_captured_tensor.h:28]:
```c++
auto deteched = JUST(tensor->detach());
auto tensor = std::shared_ptr<AutogradCapturedTensor>(new AutogradCapturedTensor(deteched));
JUST(tensor->Init());
return tensor;
```

Comment (lixinqi) [reply oneflow/core/autograd/autograd_captured_tensor.h:28]: 这样是不是去掉所有的显示CHECK和GetPtrOrThrow了呢？

File: oneflow/core/autograd/autograd_captured_tensor.h (hunk 3 above)
Comment (hjchen2) [reply oneflow/core/autograd/autograd_captured_tensor.h]: done

File: oneflow/core/autograd/autograd_captured_tensor.h (hunk 5 above)
Comment (hjchen2) [reply oneflow/core/autograd/autograd_captured_tensor.h]: done

File: oneflow/core/autograd/autograd_captured_tensor.h (hunk 6 above)
Comment (hjchen2) [reply oneflow/core/autograd/autograd_captured_tensor.h]: done

File: oneflow/core/autograd/autograd_captured_tensor.h (hunk 7 above)
Comment (hjchen2) [reply oneflow/core/autograd/autograd_captured_tensor.h]: done
